### ISSUE TITLE
Selecting text and pressing backspace deletes way more characters than desired

### ISSUE THREAD

**Issue by ghost** [autokey/autokey#196]
# ISSUE

## Classification

Bug

## Reproducibility

Always

## Summary

Whenever I have some text highlighted and I press the backspace key, instead of just deleting the selection, it also deletes 8 additional characters.

Here's the result of `autokey-gtk -l` when I perform the following actions:

1. press the keybind trigger I made: `alt+shift+j` to send `shift+left` to the current window
2. release `alt+shift+j`
3. press the backspace key
```
2018-10-09 16:20:42,442 DEBUG - iomediator - Key.ALT pressed
2018-10-09 16:20:42,538 DEBUG - iomediator - Key.SHIFT pressed
2018-10-09 16:20:43,678 DEBUG - service - Raw key: 'j', modifiers: [<Key.ALT: '<alt>'>, <Key.SHIFT: '<shift>'>], Key: J
2018-10-09 16:20:43,678 DEBUG - service - Window visible title: 'autokey-gtk -l', Window class: 'urxvt.URxvt'
2018-10-09 16:20:43,679 INFO - service - Matched hotkey phrase/script with prompt=False
2018-10-09 16:20:43,680 DEBUG - iomediator - Send via event interface
2018-10-09 16:20:43,680 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2018-10-09 16:20:43,684 DEBUG - interface - Send modified key: modifiers: ['<shift>'] key: <left>
^[[d2018-10-09 16:20:43,869 DEBUG - iomediator - Key.ALT released
2018-10-09 16:20:43,895 DEBUG - iomediator - Key.SHIFT released                                                                                                    2018-10-09 16:20:44,972 DEBUG - service - Raw key: <Key.BACKSPACE: '<backspace>'>, modifiers: [], Key: Key.BACKSPACE
2018-10-09 16:20:44,973 DEBUG - service - Window visible title: 'autokey-gtk -l', Window class: 'urxvt.URxvt'
2018-10-09 16:20:44,973 DEBUG - service - No phrase/script matched hotkey
2018-10-09 16:20:44,973 INFO - service - Undoing last abbreviation expansion
2018-10-09 16:20:44,973 DEBUG - service - Replay string: 
2018-10-09 16:20:44,974 DEBUG - service - Erase string: '<shift>+<left>'
2018-10-09 16:20:44,975 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]                                                                     2018-10-09 16:20:44,977 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]                                                                     2018-10-09 16:20:44,978 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-10-09 16:20:44,978 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-10-09 16:20:44,979 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-10-09 16:20:44,980 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-10-09 16:20:44,981 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-10-09 16:20:44,981 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]

```
## Version

AutoKey version: 0.95.3-2

Used GUI: GTK

Installed via: `yay -S autokey-py3`

Distro: Arch Linux

## Notes

I have not created any abbreviations that involve backspaces or deletions. In fact, no keybinding to my knowledge is made to send 8 backspaces. Up until now, my autokey config was working completely fine, but as soon as I updated my system via `yay -Syyu` a couple of days ago, it started having this behavior. This probably is a problem with autokey since when I kill the autokey process via `killall autokey-gtk`, the select/backspace behavior becomes normal (when I use `shift+left` and follow it up with `backspace`).

**Comment by luziferius** [comment]
Are those weird, missing line breaks in the log also in the output? Edit: You used the phrase on the terminal window running autokey, right? Those missing line breaks really puzzled me in the beginning…

Ok, I found the issue.
You are using a Phrase with `<shift>+<left>` as it’s sole content. You also have `Enable undo by pressing backspace` enabled in the settings. (`INFO - service - Undoing last abbreviation expansion`)

When you press backspace to delete the selection, you are triggering the `undo` functionality, which tries to undo the previous phrase expansion. It somehow determines that `<shift>+<left>` is 7 (or 8?) characters long and deletes the phrase. Because it is a special character combination in reality, it breaks.
That you run into this situation now is the result of this crash bug fix: https://github.com/autokey/autokey/commit/abf33455a2d2c6ae2ae848c6621f19564392340f. Now, undo reacts to phrases triggered by hotkeys without crashing the phrase expansion thread. Thus now autokey tries to undo your phrase and fails miserably.

For now, you have three options:

- Use the `delete` key on your keyboard to delete the selection
- Disable the Undo functionality in the settings (under General → Expansions)
- Use a script instead of a phrase. A simple one-liner is enough (`keyboard.send_keys("<shift>+<left>")`. Edit: I looked at your public autokey phrases. This step is infeasible…


It seems, the phrase length computation logic is broken and needs fixing. If it returns 0 in this case, the phrase should start working again.
(And the logging message quoted above needs fixing. It should state `phrase expansion`)

There is the undo functionality:
https://github.com/autokey/autokey/blob/3e60a0161e3f3d0854dac5207de2f607aa20d7a3/lib/autokey/service.py#L414-L428
It calls IoMediator.remove_string(), which looks broken. (Why does it remove len(section) characters for special keys?)
https://github.com/autokey/autokey/blob/3e60a0161e3f3d0854dac5207de2f607aa20d7a3/lib/autokey/iomediator/_iomediator.py#L168-L177

**Comment by ghost** [comment]
Wow, thank you so much! I unchecked the setting that you specified, and now it works just fine.
Your response was incredibly detailed, thorough, and timely, and I appreciate that.

P.S. Yes, I did use the phrase on the terminal window itself. Didn't feel like switching back and forth between another window haha. I just copied and pasted the entire relevant output.

**Comment by luziferius** [comment]
I thought about this problem several times, trying to figure out how to solve this.
The end goal is that you can enable undo in the settings again without breaking things.

But trying to undo phrases like the one you use proved to be quite challenging.
Basically, it boils down to something like this worst case: How do you reliably undo the phrase `abc<Alt>+<F4>123<left>`?

I think I'll implement that autokey scans the phrase content for any kind of special characters and disable undo for the particular phrase if any are found.

**Comment by luziferius** [comment]
Pushed a fix to the master branch.
In the next release, you can enable undo again in the settings.
AutoKey will undo regular phrases on `<backspace>` and disable undoing, if it contains special keys.

**Comment by josephj11** [comment]
Your fix makes sense. If you do an alt+F4, that could often do something like close a window and there would be no way AutoKey could undo that. Other keys like that would have similar, if less severe, consequences.

The remaining problems are where to document it and what, if anything, AutoKey should do if the user presses a backspace after such an expansion - expecting some sort of undo to occur. Some software beeps at you (terminal bell ...) when you try to do disallowed things, but AutoKey can't depend on having a terminal open at all and sound might be turned off or not even available.

**Comment by luziferius** [comment]
AutoKey can never know what happens on special keys.
So I used the Zen of Python as a guideline here. (Try `import this` in a Python shell) Especially this line: `In the face of ambiguity, refuse the temptation to guess.`


Currently, some limitations are documented here:
https://github.com/autokey/autokey/wiki/Known-limitations
But that page needs an update to reflect the newest changes, which I'll do later.

Previously, AutoKey did nothing on requesting undo operations, if impossible to do so. 
But impossible was defined as "not triggered by an abbreviation"...

A system beep is quite noisy, especially if you want to work in quiet environments.
Message box dialogues are annoying a.f.

Maybe a dbus based notification could work. Your Desktop environment typically already handles them in a reasonable manner, like displaying them for some seconds in a corner of the screen. Those notifications are flashy enough to notice them, but not too distracting. And they go away without needing manual confirmation.

**Comment by josephj11** [comment]
The dbus thing appears to be the path of least annoyance. (There's a law of physics about that,  isn't there? ;)  )

**Comment by luziferius** [comment]
Released version [`0.95.5`](https://github.com/autokey/autokey/releases/tag/v0.95.5), containing the bug fix.
@starun96 After upgrading to 0.95.5, you should be able to enable the undo functionality again, if you wish. AutoKey will only undo phrases not containing any special keys. Thus undo works on normal text phrases, but won’t interfere with your text navigation phrases.
